python/meterpreter/ext_server_stdapi.py

```diff
@@ -1663,6 +1663,44 @@ def stdapi_sys_process_memory_unlock(request, response):
         return error_result_windows(), response
     return ERROR_SUCCESS, response
 
+@register_function_if(has_windll)
+def stdapi_sys_process_memory_read(request, response):
+    ERROR_PARTIAL_COPY = 229
+    handle = packet_get_tlv(request, TLV_TYPE_HANDLE).get('value', 0)
+    base = packet_get_tlv(request, TLV_TYPE_BASE_ADDRESS).get('value', 0)
+    size = packet_get_tlv(request, TLV_TYPE_LENGTH).get('value', 0)
+
+    ReadProcessMemory = ctypes.windll.kernel32.ReadProcessMemory
+    ReadProcessMemory.argtypes = [ctypes.c_void_p, ctypes.c_void_p, ctypes.c_void_p, ctypes.c_size_t, ctypes.POINTER(ctypes.c_size_t)]
+    ReadProcessMemory.restype = ctypes.c_bool
+
+    buffer = ctypes.create_string_buffer(size)
+    bytes_read = ctypes.c_size_t(0)
+    if (not ReadProcessMemory(handle, base, ctypes.byref(buffer), ctypes.sizeof(buffer), ctypes.byref(bytes_read))) and (ctypes.windll.kernel32.GetLastError() != ERROR_PARTIAL_COPY):
+        return error_result_windows(), response
+
+    readed_data = buffer.raw[:bytes_read.value]
+    response += tlv_pack(TLV_TYPE_PROCESS_MEMORY, readed_data)
+    return ERROR_SUCCESS, response
+
+@register_function_if(has_windll)
+def stdapi_sys_process_memory_write(request, response):
+    ERROR_PARTIAL_COPY = 229
+    handle = packet_get_tlv(request, TLV_TYPE_HANDLE).get('value', 0)
+    base = packet_get_tlv(request, TLV_TYPE_BASE_ADDRESS).get('value', 0)
+    data = packet_get_tlv(request, TLV_TYPE_PROCESS_MEMORY).get('value', 0)
```

Comment:
This doesn't seem like a good idea. If `TLV_TYPE_PROCESS_MEMORY` isn't present, `data` will be set to 0 when it should be bytes.

`data` should always be either `None` or a `bytes` object (in Python 3 and `str` in Python 2 of course).

Really though, if any of these three parameters are not present, it should fail with `ERROR_INVALID_PARAMETER` as the Windows one does in [memory.c](https://github.com/rapid7/metasploit-payloads/blob/6bac251842e0140b78e6bd7e23e56688b83f34cb/c/meterpreter/source/extensions/stdapi/server/sys/process/memory.c#L160-L167).